File: src/jsonrpc/eth/getTransactionReceipt/eth_getTransactionReceipt.go
package eth

import (
	"encoding/json"
	"fmt"

	"github.com/ethereum/execution-apis/types"
)

// Returns the receipt of a transaction by transaction hash.
//
// Example:
// Transaction hash: "0x504ce587a65bdbdb6414a0c6c16d86a04dd79bfcc4f2950eec9634b30ce5370f"
// Result: ...
//
// Implements the eth_getTransactionReceipt JSON-RPC method.

// Method is the JSON-RPC method name
const Method = "eth_getTransactionReceipt"

// Params represents the parameters for eth_getTransactionReceipt
type Params struct {
	// 32 byte hex value
	Transaction hash types.Hash `json:"-"`
}

// MarshalJSON implements json.Marshaler for Params.
// JSON-RPC 2.0 uses positional array parameters.
func (p Params) MarshalJSON() ([]byte, error) {
	return json.Marshal([]interface{}{
		p.Transaction hash,
	})
}

// UnmarshalJSON implements json.Unmarshaler for Params.
func (p *Params) UnmarshalJSON(data []byte) error {
	var arr []json.RawMessage
	if err := json.Unmarshal(data, &arr); err != nil {
		return err
	}

	if len(arr) != 1 {
		return fmt.Errorf("expected 1 parameters, got %d", len(arr))
	}

	if err := json.Unmarshal(arr[0], &p.Transaction hash); err != nil {
		return fmt.Errorf("parameter 0 (Transaction hash): %w", err)
	}

	return nil
}

// Result represents the result for eth_getTransactionReceipt
type Result struct {
	Value types.Quantity `json:"-"`
}

// MarshalJSON implements json.Marshaler for Result.
func (r Result) MarshalJSON() ([]byte, error) {
	return json.Marshal(r.Value)
}

// UnmarshalJSON implements json.Unmarshaler for Result.
func (r *Result) UnmarshalJSON(data []byte) error {
	return json.Unmarshal(data, &r.Value)
}

File: src/jsonrpc/eth/getTransactionReceipt/eth_getTransactionReceipt.zig
const std = @import("std");
const types = @import("../../types.zig");

/// Returns the receipt of a transaction by transaction hash.
///
/// Example:
/// Transaction hash: "0x504ce587a65bdbdb6414a0c6c16d86a04dd79bfcc4f2950eec9634b30ce5370f"
/// Result: ...
///
/// Implements the `eth_getTransactionReceipt` JSON-RPC method.
pub const EthGetTransactionReceipt = @This();

/// The JSON-RPC method name
pub const method = "eth_getTransactionReceipt";

/// Parameters for `eth_getTransactionReceipt`
pub const Params = struct {
    /// 32 byte hex value
    transaction hash: types.Hash,

    pub fn jsonStringify(self: Params, jws: *std.json.Stringify) !void {
        try jws.beginArray();
        try jws.write(self.transaction hash);
        try jws.endArray();
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Params {
        if (source != .array) return error.UnexpectedToken;
        if (source.array.items.len != 1) return error.InvalidParamCount;

        return Params{
            .transaction hash = try std.json.innerParseFromValue(types.Hash, allocator, source.array.items[0], options),
        };
    }
};

/// Result for `eth_getTransactionReceipt`
pub const Result = struct {
    value: types.Quantity,

    pub fn jsonStringify(self: Result, jws: *std.json.Stringify) !void {
        try jws.write(self.value);
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Result {
        return Result{
            .value = try std.json.innerParseFromValue(types.Quantity, allocator, source, options),
        };
    }
};
